### DISCUSSION TITLE
How can I pin a specific version of CDK lib?

### DISCUSSION THREAD

**Discussion by marcioemiranda** [projen/projen#4008] [Q&A] [answered]
Hello,

I am using projen to manage the artifacts of a CDK typescript app.

In my .projenrc.ts I have the cdk lib version configured as:

cdkVersion: '2.154.0',

Running "npx projen" it generates the package.json and yarn.lock.

In the generated package.json I have:

```
...
"dependencies": {
"aws-cdk-lib": "^2.154.0",
}
...
```

However in yarn.lock I have:

```
aws-cdk-lib@^2.154.0:
  version "2.173.2"
  resolved "https://registry.yarnpkg.com/aws-cdk-lib/-/aws-cdk-lib-2.173.2.tgz#9e51332973163eb0d0a89cb67973a89b75c5f0b1"
  integrity sha512-cL9+z8Pl3VZGoO7BwdsrFAOeud/vSl3at7OvmhihbNprMN15XuFUx/rViAU5OI1m92NbV4NBzYSLbSeCwYLNyw==
  dependencies:
    "@aws-cdk/asset-awscli-v1" "^2.2.208"
    "@aws-cdk/asset-kubectl-v20" "^2.1.3"
    "@aws-cdk/asset-node-proxy-agent-v6" "^2.1.0"
    "@aws-cdk/cloud-assembly-schema" "^38.0.1"
    "@balena/dockerignore" "^1.0.2"
    case "1.6.3"
    fs-extra "^11.2.0"
    ignore "^5.3.2"
    jsonschema "^1.4.1"
    mime-types "^2.1.35"
    minimatch "^3.1.2"
    punycode "^2.3.1"
    semver "^7.6.3"
    table "^6.8.2"
    yaml "1.10.2"
```

Please notice version "2.173.2"

This is causing the following message when I run any CDK command:

```
This CDK CLI is not compatible with the CDK library used by your application. Please upgrade the CLI to the latest version.
(Cloud assembly schema version mismatch: Maximum schema version supported is 36.x.x, but found 38.0.1)
```

My CDK version is:
2.154.0 (build 0fc07f3)

In other projects where I have a proper yarn.lock, I don't get the same CLI error. For instance, here is a snippet from a project that is working properly (using the same cdk cli):

```
aws-cdk-lib@^2.154.0:
  version "2.154.0"
  resolved "https://registry.npmjs.org/aws-cdk-lib/-/aws-cdk-lib-2.154.0.tgz"
  integrity sha512-r6WmN+pk0k328Ig7yn9tqst5QR4XmyMCUG2artMgzA1YfDd4nDaP9zgJlHXIMOfVRbhgoCBX1XSSqLtN5Losgg==
  dependencies:
    "@aws-cdk/asset-awscli-v1" "^2.2.202"
    "@aws-cdk/asset-kubectl-v20" "^2.1.2"
    "@aws-cdk/asset-node-proxy-agent-v6" "^2.0.3"
    "@aws-cdk/cloud-assembly-schema" "^36.0.5"
    "@balena/dockerignore" "^1.0.2"
    case "1.6.3"
    fs-extra "^11.2.0"
    ignore "^5.3.1"
    jsonschema "^1.4.1"
    mime-types "^2.1.35"
    minimatch "^3.1.2"
    punycode "^2.3.1"
    semver "^7.6.2"
    table "^6.8.2"
    yaml "1.10.2"
``` 
Here, the version is matching the specificed cdk lib version:
version "2.154.0"

I don't know why "npx projen" is resolving to the latest cdk version, but it seems like this is causing the CLI error message.

As a side note, I've tried upgrading cdk cli to the latest version but had issues with it. My projects use aws sso to get credentials, so the following commands work fine with cdk cli 2.154.0:

aws sso login --profile <myprofile>
cdk deploy xx --profile <myprofile>

However, with the latest cdk cli it fails to get the credentials and set default CDK envs properly. It's nothing related to credentials per se, as I can run for example "aws s3 ls --profile <myprofile>" and it works. Something in the latest cdk cli is broken.

So now I am stucked since I can't use the latest cdk cli and, on the other hand, "npx projen" is generating a yarn.lock that is forcing me to upgrade the cli. What I need is to pin the CDK lib version and avoid the latest cdk cli for now.

Thanks

**Comment by mrgrain** [comment, marked as answer]
You are probably looking for

```ts
cdkVersionPinning: true
```

See: https://projen.io/docs/project-types/aws-cdk-construct-library?_highlight=cdkversionpin#depending-on-aws-cdk-lib

**Reply by marcioemiranda** [reply]
Thanks for the reply.
As a follow up comment, I was able to move on by:

- reverting package.json and yarn.lock from previous commit
- delete cdk.out and node_modules
- run yarn install
- use cdk cli normally to synth project.

**Reply by marcioemiranda** [reply]
Your suggestion solved the problem.
Now the version is pinned.
Thanks.
